api-report/tree.api.md

```diff
@@ -4,9 +4,34 @@
 
 ```ts
 
+import { Serializable } from '@fluidframework/datastore-definitions';
+
+// @public
+export interface ITreeCursor {
+    down(key: TreeKey, index: number): TreeNavigationResult;
+    keys: Iterable<TreeKey>;
+    length(key: TreeKey): number;
+    type: TreeType;
+    up(): TreeNavigationResult;
+    value: undefined | Serializable;
```

Comment:
What are the semantics of this method? Is it "if there's any way to produce a Serializable from this node, do so", or more like "if this node semantically represents a literal for a Serializable-compatible value, return that value"? If we do track character identity, for example, a String node might comprise several distinct strings in its implementation from which a single string could be generated. Or a sufficiently large String might be implemented as a balanced tree, and converting that to a single string might be prohibitively expensive and iteration recommended instead. We've also talked about value on complex nodes return either a JSON string or an object tree that is trivially JSON-serializable. Could we live without this method in this interface, and have more strongly-typed methods in individual schemas? Finally, how does this property relate to Object.valueOf()?

Reply:
'value' is how the schema/api layers will read the raw node values.  There's a fair question as to whether the returned value should be deserialized, or if we should return the raw bytes from the chunk/changeset.

Interesting question about large node values.  Having specialized accessors for different NodeTypes instead of a single '.value' property could be a win.

Reply:
I would certainly encourage specialized accessors instead of a generic property. Even if we do ultimately decide to factor out some common patterns, I think we'll end up in a better place if we approach the problem from that direction, rather than assuming a generic property has, um, value and trying to make that work in as many cases as possible. I think we're going to need specialized mutation methods, anyway. For example, if we have different primitives for INT16 and UINT32, those will both probably be implemented with JS numbers, but will have different checks in their set methods.

Reply:
Per a later comment, we could refine that as an API that allows you to efficiently consume adjacent segments of homogenous types (e.g., a run of characters in a string) at whatever granularity is convenient for the tree implementation.

For a run of length 1, it provides a 'T'.  For 2 or more it provides a string (for codepoints) or T[].